packages/common/async/src/mutex.ts

```diff
@@ -133,3 +140,31 @@ export const synchronized = (
   };
   Object.defineProperty(descriptor.value, 'name', { value: propertyName + '$synchronized' });
 };
+
+export const acquireInContext = async (
+  ctx: Context,
+  mutex: Mutex,
+  options: { releaseOnDispose: boolean } = { releaseOnDispose: true },
+): Promise<MutexGuard> => {
+  let guard: MutexGuard | undefined;
+  try {
+    return await cancelWithContext(
+      ctx,
+      (async () => {
+        guard = await mutex.acquire();
+        if (ctx.disposed) {
+          guard.release();
+          guard = undefined;
+          throw new Error('Guard released because acquired after context was disposed.');
+        }
+        if (options.releaseOnDispose) {
+          ctx.onDispose(() => guard?.release());
```

Comment:
onDispose returns a function to clear this callback. Ideally we'd call it when the guard gets released. Otherwise the callbacks can pile-up in a long-running context